File: CodeArchieve/UOJ/Test-Round-1/83-水题出题人/problem_83/bogo_sort.pas
program bogo_sort;

{
	请Pascal选手注意，这里使用了自然溢出
	用命令行fpc code.pas方式直接编译在运行时不会报错
	如果用IDE默认的编译选项有可能报错

	For Pascal programmers, please note that we use natural integer overflow here
	If you compile this code with "fpc code.pas" in command line, it won't get a Runtime Error
	If you cmopile with IDE, Runtime Error may occur
}

var counter, n :longint;
var a :array[1..1000000] of longint;

{
	下面实现了一个伪随机数生成器

	Below is a random number generator
}

var seed, RNG_a, RNG_b :longword;

procedure RNG_init;
begin
	inc(counter);
	seed := 2166136261;
	inc(counter);
	RNG_a := 26648577;
	inc(counter);
	RNG_b := 10108929;
end;

function RNG_next :longint;
begin
	inc(counter);
	seed := seed * RNG_a + RNG_b;
	inc(counter);
	exit(seed and $7fffffff);
end;

function check :boolean;
var i: longint;
begin
	inc(counter);
	for i := 2 to n do begin
		inc(counter);
		if a[i - 1] > a[i] then begin
			inc(counter);
			exit(false);
		end;
	end;
	inc(counter);
	exit(true);
end;

procedure bogo_sort_main;
var i, j, temp :longint;
begin
	inc(counter);
	readln(n);
	inc(counter);
	for i := 1 to n do begin
		inc(counter);
		readln(a[i]);
	end;
	inc(counter);
	RNG_init;
	inc(counter);
	while not check do begin
		inc(counter);
		for i := 1 to n do begin
			inc(counter);
			seed := (seed * 16777619) xor a[i];
		end;
		inc(counter);
		for i := 1 to n do begin
			inc(counter);
			j := RNG_next mod n + 1;
			inc(counter);
			temp := a[i];
			inc(counter);
			a[i] := a[j];
			inc(counter);
			a[j] := temp;
		end;
		inc(counter);
		RNG_a := (RNG_next shr 5) + 1;
		inc(counter);
		RNG_b := (RNG_next shr 5) + 3;
	end;
	inc(counter);
	for i := 1 to n do begin
		inc(counter);
		write(a[i], ' ');
	end;
end;

begin
	counter := 0;
	bogo_sort_main;
	writeln;
	if counter <= 2000000 then begin
		writeln('Accepted! The counter value is :', counter);
	end else begin
		writeln('Time Limit Exceeded because the counter value is :', counter);
	end;
end.

File: CodeArchieve/UOJ/Test-Round-1/83-水题出题人/problem_83/bogo_sort.cpp
#include <stdio.h>

int counter, n;
int a[1000001];

// 下面实现了一个伪随机数生成器

// Below is a random number generator

unsigned int seed, RNG_a, RNG_b;

void RNG_init(){
	counter++;
	seed = 2166136261u;
	counter++;
	RNG_a = 26648577u;
	counter++;
	RNG_b = 10108929u;
}

int RNG_next(){
	counter++;
	seed = seed * RNG_a + RNG_b;
	counter++;
	return seed & 0x7fffffff;
}

bool check(){
	int i;
	counter++;
	for(i = 2; i <= n; i++){
		counter++;
		if(a[i - 1] > a[i]){
			counter++;
			return false;
		}
	}
	counter++;
	return true;
}

void bogo_sort_main(){
	int i, j, temp;
	counter++;
	scanf("%d", &n);
	counter++;
	for(i = 1; i <= n; i++){
		counter++;
		scanf("%d", &a[i]);
	}
	counter++;
	RNG_init();
	counter++;
	while(!check()){
		counter++;
		for(i = 1; i <= n; i++){
			counter++;
			seed = (seed * 16777619u) ^ a[i];
		}
		counter++;
		for(i = 1; i <= n; i++){
			counter++;
			j = RNG_next() % n + 1;
			counter++;
			temp = a[i];
			counter++;
			a[i] = a[j];
			counter++;
			a[j] = temp;
		}
		counter++;
		RNG_a = (RNG_next() >> 5) + 1;
		counter++;
		RNG_b = (RNG_next() >> 5) + 3;
	}
	counter++;
	for(i = 1; i <= n; i++){
		counter++;
		printf("%d ", a[i]);
	}
}

int main(){
	counter = 0;
	bogo_sort_main();
	printf("\n");
	if(counter <= 2000000){
		printf("Accepted! The counter value is :%d\n", counter);
	}else{
		printf("Time Limit Exceeded because the counter value is :%d\n", counter);
	}
	return 0;
}
